src/psyclone/f2pygen.py

```diff
@@ -766,62 +765,148 @@ def __init__(self, parent, content):
             myline = reader.next()
             self._decl = fparser1.typedecl_statements.Integer(parent.root,
                                                               myline)
-        elif datatype.lower() == "real":
+        elif dtype == "real":
             reader = FortranStringReader("real :: vanilla")
             reader.set_format(fort_fmt)
             myline = reader.next()
             self._decl = fparser1.typedecl_statements.Real(parent.root, myline)
+        elif dtype == "logical":
+            reader = FortranStringReader("logical :: vanilla")
+            reader.set_format(fort_fmt)
+            myline = reader.next()
+            self._decl = fparser1.typedecl_statements.Logical(parent.root,
+                                                              myline)
         else:
-            raise RuntimeError(
-                "f2pygen:DeclGen:init: Only integer and real are currently"
-                " supported and you specified '{0}'".format(datatype))
-        # make a copy of entity_decls as we may modify it
+            # Defensive programming in case SUPPORTED_TYPES is added to
+            # but not handled here
+            raise NotImplementedError(
+                "Internal error: type '{0}' is in DeclGen.SUPPORTED_TYPES "
+                "but not handled by constructor.".format(dtype))
+
+        # Make a copy of entity_decls as we may modify it
         local_entity_decls = entity_decls[:]
-        self._decl.entity_decls = local_entity_decls
+        if initial_values:
+            # Create a list of 2-tuples
+            value_pairs = zip(local_entity_decls, initial_values)
+            # Construct an assignment from each tuple
+            self._decl.entity_decls = ["=".join(_) for _ in value_pairs]
+        else:
+            self._decl.entity_decls = local_entity_decls
+
         my_attrspec = []
         if intent != "":
             my_attrspec.append("intent({0})".format(intent))
-        if pointer is not False:
+        if pointer:
             my_attrspec.append("pointer")
-        if allocatable is not False:
+        if allocatable:
             my_attrspec.append("allocatable")
+        if save:
+            my_attrspec.append("save")
         self._decl.attrspec = my_attrspec
         if dimension != "":
             my_attrspec.append("dimension({0})".format(dimension))
         if kind is not "":
             self._decl.selector = ('', kind)
         BaseGen.__init__(self, parent, self._decl)
 
+    @staticmethod
+    def _check_initial_values(dtype, values):
+        '''
+        Check that the supplied values are consistent with the requested
+        data type.
+        :param str dtype: Fortran intrinsic type
+        :param list values: list of values as strings
+        :raises RuntimeError: if the supplied values are not consistent
+                              with the specified data type
+        '''
+        from fparser.two.pattern_tools import abs_name, \
```

Comment:
This is something a "full" parser would be able to check.

Reply:
Good plan, I shall do that.